aspnetcore/client-side/dotnet-interop/index.md

```diff
@@ -0,0 +1,948 @@
+    [JSImport("PromisesShim.Wait2Seconds", "PromisesShim")]
+    public static partial Task Wait2Seconds();
+
+    [JSImport("PromisesShim.WaitGetString", "PromisesShim")]
+    public static partial Task<string> WaitGetString();
+
+    // Some return types require a [return: JSMarshalAs...] declaring the
+    // Promise's return type corresponding to Task<T>.
+    [JSImport("PromisesShim.WaitGetDate", "PromisesShim")]
+    [return: JSMarshalAs<JSType.Promise<JSType.Date>>()]
+    public static partial Task<DateTime> WaitGetDate();
+
+    [JSImport("PromisesShim.FetchCurrentUrl", "PromisesShim")]
+    public static partial Task<string> FetchCurrentUrl();
+
+    [JSImport("PromisesShim.AsyncFunction", "PromisesShim")]
+    public static partial Task AsyncFunction();
+
+    [JSImport("PromisesShim.ConditionalSuccess", "PromisesShim")]
+    public static partial Task ConditionalSuccess(bool shouldSucceed);
+}
+
+public static class PromisesUsage
+{
+    public static async Task Run()
+    {
+        await JSHost.ImportAsync("PromisesShim", "/PromisesShim.js");
+
+        Stopwatch sw = new Stopwatch();
+        sw.Start();
+
+        await PromisesInterop.Wait2Seconds(); // Await Promise
+        Console.WriteLine($"Waited {sw.Elapsed.TotalSeconds:#.0}s.");
+
+        sw.Restart();
+        string str = 
+            await PromisesInterop.WaitGetString(); // Await promise (string return)
+        Console.WriteLine(
+            $"Waited {sw.Elapsed.TotalSeconds:#.0}s for WaitGetString: '{str}'");
+
+        sw.Restart();
+        // Await promise with string return.
+        DateTime date = await PromisesInterop.WaitGetDate();
+        Console.WriteLine(
+            $"Waited {sw.Elapsed.TotalSeconds:#.0}s for WaitGetDate: '{date}'");
+
+        // Await a JS fetch.
+        string responseText = await PromisesInterop.FetchCurrentUrl();
+        Console.WriteLine($"responseText.Length: {responseText.Length}");
+
+        sw.Restart();
+
+        await PromisesInterop.AsyncFunction(); // Await an async JS method
+        Console.WriteLine(
+            $"Waited {sw.Elapsed.TotalSeconds:#.0}s for AsyncFunction.");
+
+        try {
+            // Handle a promise rejection. Await an async JS method.
+            await PromisesInterop.ConditionalSuccess(shouldSucceed: false);
+        }
+        catch(JSException ex) // Catch JS exception
+        {
+            Console.WriteLine($"JS Exception Caught: '{ex.Message}'");
+        }
+    }
+}
+```
+
+The preceding example displays the following output in the browser's debug console:
+
+> :::no-loc text="Waited 2.0s.":::  
+> :::no-loc text="Waited .5s for WaitGetString: 'String From Resolve'":::  
+> :::no-loc text="Waited .5s for WaitGetDate: '11/24/1988 12:00:00 AM'":::  
+> :::no-loc text="responseText.Length: 582":::  
+> :::no-loc text="Waited 2.0s for AsyncFunction.":::  
+> :::no-loc text="JS Exception Caught: 'Reject: ShouldSucceed == false'":::
+
+## Subscribing to JS events
```

Comment:
somewhere we should mention that interacting with individual properties of DOM via JS interop will be slow and create lot of proxies (GC pressure). This pattern it not recommended. I'm voting for not including it in the official docs, it could be on somebody's blog instead.

Reply:
Now I see the perf section below. Maybe we can put the perf section first and keep this DOM section after it. With a note that dealing with more than handful of elements via JS interop is bad pattern.